### ISSUE TITLE
Add Train-dev set and train-dev_step() method for error analysis 

### ISSUE THREAD

**Issue by MinWang1997** [PyTorchLightning/pytorch-lightning#10539]
## 🚀 Feature

Add Train-dev set and train-dev_step() method for error analysis 

### Motivation

[CS230](https://cs230.stanford.edu/blog/split/) suggests to have additional train-dev dataset to compare the error with validation set, so that we can know the bias of our model.

### Pitch

It would be nice if you also add Train-dev set for datamodule setup and train-dev_step() method for error analysis. 

### Alternatives

Or could you tell me how to use current api to achieve it?
Since I pass lightning datamodule to trainer.validate(), I can only use validation dataset.

### Additional context

![34C82B63-4037-4B8A-917F-D02C7AE2F931](https://user-images.githubusercontent.com/75472853/141769757-137fba53-418e-4fd1-9e62-e0f7fdbd9ff6.jpeg)
![73389C83-4A97-4359-A855-84B870331E1B](https://user-images.githubusercontent.com/75472853/141769948-e6cefe3b-e1a6-4ed0-80d3-ee0531eccf6c.png)

**Comment by rohitgr7** [comment]
that is what validation is for exactly, to evaluate your performance on a dataset on which the model isn't trained yet and how your model is performing while training. We provided the `.validate` method to just run the validation loop over the val dataset if the user wants to. To find the best model out of all training/research, we have `.test` for it.

**Comment by MinWang1997** [comment]
> that is what validation is for exactly, to evaluate your performance on a dataset on which the model isn't trained yet and how your model is performing while training. We provided the `.validate` method to just run the validation loop over the val dataset if the user wants to. To find the best model out of all training/research, we have `.test` for it.

I mean now he suggested to split dataset into train, train-dev, dev(validation ) and test set (4 subsets in total.) to solve the problem of data distribution dismatch.
He suggests to train model on your training set and validate on Val (dev) set, also validate on train-dev set to see the error/accuracy difference.

We have already 3 methods for train/Val/test_step.
It would be nice if we add additional train-val(train-dev)_step for trainer and corresponding datamodule setup().


Now, it is not possible to validate another set inside datamodule(except for Val set).
I hope that I can split datasets into 4 splits in datamodule setup() and then validate my model on train-dev set and also validation set to see the difference.

**Comment by MinWang1997** [comment]
concretely, 

I split my training data name list intrain_dev_split() and train_list, train_dev_lst, dev_lst 

     ```
   train_list, train_dev_lst, dev_lst = train_dev_split(self.seed)

        # for training data
        if dataset_name == 'train':
            self.add_noise = True

            self.all_lst = train_list  # get all the training data index

            self.minibatch_per_epoch = len(self.all_lst) // config.mini_batch_size
            self.real_lst = []
            self.syn_lst = []
            for item in self.all_lst:  # separate real data and synthesis data of training dataset to 2 lists
                if item[:5] == 'data/':  # real
                    self.real_lst.append(item)
                else:
                    self.syn_lst.append(item)  # syn: prefix : data_syn/

                # validation set
        if dataset_name == 'val':
            self.pp_data = None  #???
            self.add_noise = False
            self.all_lst = dev_lst

        **# train-dev set
        if dataset_name == 'train_dev':
            self.pp_data = None  #???
            self.add_noise = False
            self.all_lst = train_dev_lst**

        # for test set
        if dataset_name == 'test':
            self.pp_data = None  #???
            self.add_noise = False
            self.path = 'datasets/ycb/dataset_config/test_data_list.txt'
            self.all_lst = bs_utils.read_lines(self.path) 
```
in datamodule I defined them as follows:
```
  def setup(self, stage: Optional[str] = None):
        """Split the train and valid dataset."""
        if stage in (None, "fit", "validate"):
            # init datasets
            self.dataset_train = dataset_desc.Dataset('train')
            self.dataset_val = dataset_desc.Dataset('val')
            **self.dataset_train_dev = dataset_desc.Dataset('train_dev')**

        if stage in(None, "test", "predict"):
            self.dataset_test = dataset_desc.Dataset('test')
```
I hope that I can run different setting for trainer like this:

```
   if args.trainer == "fit":  # train your model
        # tune to find the learning rate
        # trainer.tune(model, datamodule=dm)
        # model.learning_rate
        trainer.fit(model, datamodule=dm)

    if args.trainer == "val":  # validate your model separately from fit if needed
        trainer.validate( datamodule=dm, ckpt_path=args.checkpoint)

    **if args.trainer == "train-dev":  # validate your model separately from fit if needed
        trainer.validate/train-dev( datamodule=dm, ckpt_path=args.checkpoint)**

    if args.trainer == "test":  # test before you need to write your paper
        trainer.test(datamodule=dm, ckpt_path=args.checkpoint)

    if args.trainer == "predict":  # inference
        trainer.predict(datamodule=dm, ckpt_path=args.checkpoint)
```
